File: sdks/typescript/src/budget.ts
import { InvalidBudget } from "./errors.js";
import { Ledger } from "./ledger.js";
import { CircuitBreaker } from "./circuit_breaker.js";
import { BudgetSession, SessionOptions, Report } from "./session.js";

export interface BudgetOptions {
  /** Fraction of budget at which the soft-limit callback fires. Default: 0.9 */
  softLimit?: number;
  /** Max repeated calls in the loop window before LoopDetected is thrown. Default: 10 */
  loopMaxCalls?: number;
  /** Loop detection window in seconds. Default: 60 */
  loopWindowSeconds?: number;
  /** Fraction of budget to hold back from hard limit enforcement. Default: 0 */
  finalizationReserve?: number;
  /** Called once when spending exceeds the soft limit. */
  onSoftLimit?: ((report: Report) => void) | undefined;
  /** Called when BudgetExhausted is thrown. */
  onHardLimit?: ((report: Report) => void) | undefined;
  /** Called when LoopDetected is thrown. */
  onLoopDetected?: ((report: Report) => void) | undefined;
}

/**
 * AgentBudget is the entry point. Create one per agent type, then call
 * `newSession()` for each agent run.
 *
 * @example
 * ```ts
 * import { AgentBudget } from "agentbudget";
 *
 * const budget = new AgentBudget("$5.00");
 * const session = budget.newSession();
 *
 * const resp = await openai.chat.completions.create({ model: "gpt-4o", ... });
 * session.wrapOpenAI(resp);
 *
 * console.log(session.spent, session.remaining);
 * session.close();
 * ```
 */
export class AgentBudget {
  private readonly _budget: number;
  private readonly _reserveBudget: number;
  private readonly _softLimit: number;
  private readonly _loopMaxCalls: number;
  private readonly _loopWindowSeconds: number;
  private readonly _onSoftLimit?: ((r: Report) => void) | undefined;
  private readonly _onHardLimit?: ((r: Report) => void) | undefined;
  private readonly _onLoopDetected?: ((r: Report) => void) | undefined;

  constructor(maxSpend: string | number, opts: BudgetOptions = {}) {
    const raw = parseBudget(maxSpend);

    const finReserve = opts.finalizationReserve ?? 0;
    if (finReserve < 0 || finReserve >= 1) {
      throw new InvalidBudget(`finalizationReserve must be in [0, 1), got ${finReserve}`);
    }

    this._reserveBudget = raw;
    this._budget = raw * (1 - finReserve);
    this._softLimit = opts.softLimit ?? 0.9;
    this._loopMaxCalls = opts.loopMaxCalls ?? 10;
    this._loopWindowSeconds = opts.loopWindowSeconds ?? 60;
    this._onSoftLimit = opts.onSoftLimit;
    this._onHardLimit = opts.onHardLimit;
    this._onLoopDetected = opts.onLoopDetected;
  }

  /** Total configured budget including any finalization reserve. */
  get maxSpend(): number { return this._reserveBudget; }

  /** Create a new budget-enforced session for an agent run. */
  newSession(opts?: SessionOptions): BudgetSession {
    const ledger = new Ledger(this._budget);
    const breaker = new CircuitBreaker(
      this._softLimit,
      this._loopMaxCalls,
      this._loopWindowSeconds
    );
    return new BudgetSession(ledger, breaker, {
      onSoftLimit: this._onSoftLimit,
      onHardLimit: this._onHardLimit,
      onLoopDetected: this._onLoopDetected,
      ...opts,
    });
  }
}

function parseBudget(value: string | number): number {
  if (typeof value === "number") {
    if (value <= 0) throw new InvalidBudget(String(value));
    return value;
  }
  const cleaned = value.trim().replace(/^\$/, "").trim();
  const amount = parseFloat(cleaned);
  if (isNaN(amount) || amount <= 0) throw new InvalidBudget(value);
  return amount;
}

File: sdks/go/budget.go
package agentbudget

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"sync/atomic"
	"time"
)

// Budget is the entry point for AgentBudget. Create one per application or agent type,
// then call NewSession() for each agent run.
type Budget struct {
	budget         float64 // enforced ceiling (after finalization reserve)
	reserveBudget  float64 // full budget including reserve
	softLimit      float64
	loopMaxCalls   int
	loopWindowSecs float64
	onSoftLimit    func(Report)
	onHardLimit    func(Report)
	onLoopDetected func(Report)
}

// Option configures a Budget.
type Option func(*budgetConfig)

type budgetConfig struct {
	softLimit      float64
	loopMaxCalls   int
	loopWindowSecs float64
	onSoftLimit    func(Report)
	onHardLimit    func(Report)
	onLoopDetected func(Report)
	finReserve     float64
}

// WithSoftLimit sets the fraction of budget at which the soft-limit callback fires.
// Default: 0.9 (90%).
func WithSoftLimit(fraction float64) Option {
	return func(c *budgetConfig) { c.softLimit = fraction }
}

// WithLoopDetection configures loop detection. maxCalls is the max number of repeated
// calls in window seconds before LoopDetected is raised.
func WithLoopDetection(maxCalls int, windowSeconds float64) Option {
	return func(c *budgetConfig) {
		c.loopMaxCalls = maxCalls
		c.loopWindowSecs = windowSeconds
	}
}

// WithOnSoftLimit sets a callback invoked once when spending exceeds the soft limit.
func WithOnSoftLimit(fn func(Report)) Option {
	return func(c *budgetConfig) { c.onSoftLimit = fn }
}

// WithOnHardLimit sets a callback invoked when BudgetExhausted is raised.
func WithOnHardLimit(fn func(Report)) Option {
	return func(c *budgetConfig) { c.onHardLimit = fn }
}

// WithOnLoopDetected sets a callback invoked when LoopDetected is raised.
func WithOnLoopDetected(fn func(Report)) Option {
	return func(c *budgetConfig) { c.onLoopDetected = fn }
}

// WithFinalizationReserve reserves a fraction of the budget for a final step.
// E.g. WithFinalizationReserve(0.05) on a $1.00 budget fires the hard limit at $0.95.
func WithFinalizationReserve(fraction float64) Option {
	return func(c *budgetConfig) { c.finReserve = fraction }
}

// New creates a Budget with the given maximum spend.
// maxSpend may be a float64, int, or a string like "$5.00" or "5.00".
func New(maxSpend any, opts ...Option) (*Budget, error) {
	raw, err := parseBudget(maxSpend)
	if err != nil {
		return nil, err
	}

	cfg := &budgetConfig{
		softLimit:      0.9,
		loopMaxCalls:   10,
		loopWindowSecs: 60.0,
	}
	for _, o := range opts {
		o(cfg)
	}
	if cfg.finReserve < 0 || cfg.finReserve >= 1 {
		return nil, fmt.Errorf("finalization reserve must be in [0, 1), got %.2f", cfg.finReserve)
	}

	enforced := raw * (1.0 - cfg.finReserve)

	return &Budget{
		budget:         enforced,
		reserveBudget:  raw,
		softLimit:      cfg.softLimit,
		loopMaxCalls:   cfg.loopMaxCalls,
		loopWindowSecs: cfg.loopWindowSecs,
		onSoftLimit:    cfg.onSoftLimit,
		onHardLimit:    cfg.onHardLimit,
		onLoopDetected: cfg.onLoopDetected,
	}, nil
}

// MaxSpend returns the total configured budget including any finalization reserve.
func (b *Budget) MaxSpend() float64 { return b.reserveBudget }

// SessionOption configures a Session.
type SessionOption func(*sessionConfig)

type sessionConfig struct {
	id string
}

// WithSessionID sets a custom session identifier.
func WithSessionID(id string) SessionOption {
	return func(c *sessionConfig) { c.id = id }
}

var sessionCounter uint64

// NewSession creates a new budget-enforced session.
// Each call to the agent should use its own session.
func (b *Budget) NewSession(opts ...SessionOption) *Session {
	cfg := &sessionConfig{}
	for _, o := range opts {
		o(cfg)
	}
	if cfg.id == "" {
		n := atomic.AddUint64(&sessionCounter, 1)
		cfg.id = fmt.Sprintf("sess_%d_%d", time.Now().UnixNano(), n)
	}

	ledger := newLedger(b.budget)
	breaker := newCircuitBreaker(b.softLimit, b.loopMaxCalls, b.loopWindowSecs)
	return newSession(cfg.id, ledger, breaker, b.onSoftLimit, b.onHardLimit, b.onLoopDetected)
}

// parseBudget converts various types to a float64 budget.
func parseBudget(value any) (float64, error) {
	switch v := value.(type) {
	case float64:
		if v <= 0 {
			return 0, &InvalidBudget{Value: fmt.Sprintf("%v", v)}
		}
		return v, nil
	case float32:
		return parseBudget(float64(v))
	case int:
		return parseBudget(float64(v))
	case int64:
		return parseBudget(float64(v))
	case string:
		cleaned := strings.TrimSpace(strings.TrimPrefix(strings.TrimSpace(v), "$"))
		amount, err := strconv.ParseFloat(cleaned, 64)
		if err != nil || amount <= 0 {
			return 0, &InvalidBudget{Value: v}
		}
		return amount, nil
	default:
		return 0, &InvalidBudget{Value: fmt.Sprintf("%v", value)}
	}
}

// round6 rounds to 6 decimal places (defined in ledger.go, referenced here for clarity).
var _ = math.Round // ensure math imported
